crates/store/re_protos/proto/rerun/v0/catalog.proto

```diff
@@ -0,0 +1,249 @@
+    //
+    // This API is only supported for some catalog entries.
+    rpc RegisterResources(RegisterResourcesRequest) returns (rerun.common.v0.DataframePart) {}
+
+    // Remove a resource from the a manually-managed catalog entry
+    //
+    // This API is only supported for some catalog entries.
+    rpc UnregisterResource(UnregisterResourcesRequest) returns (UnregisterResourceResponse) {}
+
+    // Remove all resources from a manually-managed catalog entry
+    //
+    // This API is only supported for some catalog entries.
+    rpc UnregisterAllResources(UnregisterAllResourcesRequest)
+        returns (UnregisterAllResourcesResponse) {}
+
+    // Create an index for a specific column in the catalog entry
+    rpc CreateIndex(CreateIndexRequest) returns (CreateIndexResponse) {}
+
+    // Search over an existing index  in the catalog entry.
+    //
+    // The response to `SearchIndex` a RecordBatch with 3 columns:
+    // - 'resource_id' column with the id of the resource
+    // - timepoint column with the values representing the points in time
+    // where index query matches. What time points are matched depends on the type of
+    // index that is queried. For example for vector search it might be timepoints where
+    // top-K matches are found within *each* resource in the indexed entry. For inverted index
+    // it might be timepoints where the query string is found in the indexed column
+    // - 'data' column with the data that is returned for the matched timepoints
+    rpc SearchIndex(SearchIndexRequest) returns (stream rerun.common.v0.DataframePart) {}
+
+    // ---------------- Resource APIs ------------------
+    // These APIs do not need to be mediated by a catalog entry
+    // The resource-id is sufficient to uniquely locate the resource within the catalog.
+
+    rpc GetResourceSchema(GetResourceSchemaRequest) returns (GetResourceSchemaResponse) {}
+
+    rpc ListChunks(ListChunksRequest) returns (stream ChunkInfo) {}
+
+    rpc GetChunks(stream GetChunkRequest) returns (stream rerun.common.v0.RerunChunk) {}
+
+    // Run a Query against a single resource in the storage node
+    rpc ResourceQuery(ResourceQueryRequest) returns (stream rerun.common.v0.DataframePart) {}
+}
+
+message CatalogEntry {
+    string name = 1;
+}
+
+// ---------------- Catalog APIs ------------------
+
+// TODO
+message ListEntriesRequest {}
+message ListEntriesResponse {}
+
+// TODO
+message AddEntryRequest {}
+message AddEntryResponse {}
+
+// TODO
+message RemoveEntryRequest {}
+message RemoveEntryResponse {}
+
+// ---------------- Collection APIs ------------------
+
+message GetPropertyTableSchemaRequest {
+    CatalogEntry entry = 1;
+}
+
+message GetPropertyTableSchemaResponse {
+    // The schema of the property table for the given catalog entry.
+    rerun.common.v0.Schema schema = 1;
+}
+
+message UpdateResourcePropertiesRequest {
+    CatalogEntry entry = 1;
+
+    // This dataframe must contain a column indicating the resource_id
```

Comment:
It would be really great to have that information as actual types. Conceptually, it would be great to have `struct Dataframe<I: Index>`, where `Index`, in this case, would be something like `ResourceId`. For data frames that don't have an explicit index (the order is the index), we could default to `struct Dataframe<I: Index = ()>`.

Of course `proto` does not give us generics, but maybe we could "monomorphize" the few instances where this is the case. We could then have specialized constructors + deserializers that ensure these invariants.